File: R/fnorm.R
#' Folded Normal Distribution
#'
#' Density, cumulative distribution, quantile functions and random number
#' generation for the folded normal distribution with the location
#' parameter mu and the scale sigma (which corresponds to standard
#' deviation in normal distribution).
#'
#' The distribution has the following density function:
#'
#' f(x) = 1/sqrt(2 pi) (exp(-(x-mu)^2 / (2 sigma^2)) + exp(-(x+mu)^2 / (2 sigma^2)))
#'
#' Both \code{pfnorm} and \code{qfnorm} are returned for the lower
#' tail of the distribution.
#'
#' @template author
#' @keywords distribution
#'
#' @param q vector of quantiles.
#' @param p vector of probabilities.
#' @param n number of observations. Should be a single number.
#' @param mu vector of location parameters (means).
#' @param sigma vector of scale parameters.
#' @param log if \code{TRUE}, then probabilities are returned in
#' logarithms.
#'
#' @return Depending on the function, various things are returned
#' (usually either vector or scalar):
#' \itemize{
#' \item \code{dfnorm} returns the density function value for the
#' provided parameters.
#' \item \code{pfnorm} returns the value of the cumulative function
#' for the provided parameters.
#' \item \code{qfnorm} returns quantiles of the distribution. Depending
#' on what was provided in \code{p}, \code{mu} and \code{sigma}, this
#' can be either a vector or a matrix, or an array.
#' \item \code{rfnorm} returns a vector of random variables
#' generated from the fnorm distribution. Depending on what was
#' provided in \code{mu} and \code{sigma}, this can be either a vector
#' or a matrix or an array.
#' }
#'
#' @examples
#' x <- dfnorm(c(-1000:1000)/200, 0, 1)
#' plot(x, type="l")
#'
#' x <- pfnorm(c(-1000:1000)/200, 0, 1)
#' plot(x, type="l")
#'
#' qfnorm(c(0.025,0.975), 0, c(1,2))
#'
#' x <- rfnorm(1000, 0, 1)
#' hist(x)
#'
#' @references \itemize{
#' \item Wikipedia page on folded normal distribution:
#' \url{https://en.wikipedia.org/wiki/Folded_normal_distribution}.
#' }
#'
#' @rdname fnorm-distribution
#' @importFrom stats optim pnorm qnorm rnorm
#' @export dfnorm
dfnorm <- function(q, mu=0, sigma=1, log=FALSE){
    fnormReturn <- 1/(sqrt(2 * pi)*sigma) * (exp(-(q-mu)^2 / (2 * sigma^2)) +
                                                 exp(-(q+mu)^2 / (2 * sigma^2)));
    if(log){
        fnormReturn[] <- log(fnormReturn);
    }
    fnormReturn[q<0] <- 0;

    return(fnormReturn);
}

#' @rdname fnorm-distribution
#' @export pfnorm
#' @aliases pfnorm
pfnorm <- function(q, mu=0, sigma=1){
    fnormReturn <- pnorm(q, mu, sigma) + pnorm(q, -mu, sigma) - 1;
    fnormReturn[q<0] <- 0;
    return(fnormReturn);
}

#' @rdname fnorm-distribution
#' @export qfnorm
#' @aliases qfnorm
qfnorm <- function(p, mu=0, sigma=1){
    # p <- unique(p);
    # mu <- unique(mu);
    # sigma <- unique(sigma);
    CF <- function(A, P, mu, sigma){
        probability <- pfnorm(A, mu, sigma);
        return((P-probability)^2);
    }

    lengthMax <- max(length(p),length(mu),length(sigma));
    # If length of p, mu and sigma differs, then go difficult. Otherwise do simple stuff
    if(any(!c(length(p),length(mu),length(sigma)) %in% c(lengthMax, 1))){
        fnormReturn <- array(0,c(length(p),length(mu),length(sigma)),
                             dimnames=list(paste0("p=",p),paste0("mu=",mu),paste0("sigma=",sigma)));
        fnormReturn[p==0,,] <- 1;
        fnormReturn[p==1,,] <- Inf;
        probsToEstimate <- which(fnormReturn[,1,1]==0);
        fnormReturn[fnormReturn==1] <- 0;

        for(j in probsToEstimate){
            for(k in 1:length(sigma)){
                for(i in 1:length(mu)){
                    fnormReturn[j,i,k] <- optim(abs(qnorm(p[j],mu[i],sigma[k])), CF, method="L-BFGS-B", lower=0,
                                                P=p[j], mu=mu[i], sigma=sigma[k])$par;
                }
            }
        }

        # Drop the redundant dimensions
        fnormReturn <- fnormReturn[,,];
    }
    else{
        fnormReturn <- rep(0,lengthMax);
        p <- rep(p,lengthMax)[1:lengthMax];
        mu <- rep(mu,lengthMax)[1:lengthMax];
        sigma <- rep(sigma,lengthMax)[1:lengthMax];

        for(i in 1:lengthMax){
            if(p[i]==0){
                fnormReturn[i] <- 0;
            }
            else if(p[i]==1){
                fnormReturn[i] <- Inf;
            }
            else{
                fnormReturn[i] <- optim(abs(qnorm(p[i],mu[i],sigma[i])), CF, method="L-BFGS-B", lower=0,
                                        P=p[i], mu=mu[i], sigma=sigma[i])$par;
            }
        }
    }
    return(fnormReturn);
}

#' @rdname fnorm-distribution
#' @export rfnorm
#' @aliases rfnorm
rfnorm <- function(n=1, mu=0, sigma=1){
    fnormReturn <- abs(rnorm(n,mu,sigma));
    return(fnormReturn);
}

File: python/src/greybox/distributions/fnorm.py
"""Folded Normal distribution functions.

Density, cumulative distribution, quantile functions and random number
generation for the folded normal distribution.
"""

import numpy as np
from scipy import stats
from scipy import optimize


def dfnorm(q, loc=0, scale=1, log=False):
    """Folded Normal distribution density.

    f(x) = 1/sqrt(2*pi*scale^2) * (
        exp(-(x-loc)^2 / (2*scale^2)) + exp(-(x+loc)^2 / (2*scale^2)))

    Parameters
    ----------
    q : array_like
        Quantiles.
    loc : float
        Location parameter.
    scale : float
        Scale parameter.
    log : bool
        If True, return log-density.

    Returns
    -------
    array
        Density values.
    """
    q = np.asarray(q)
    abs_q = np.abs(q)
    if log:
        # log(f(x)) = log(exp(a) + exp(b)) - log(sqrt(2*pi)*scale)
        # where a = -(|x|-loc)^2/(2*scale^2), b = -(|x|+loc)^2/(2*scale^2)
        a = -((abs_q - loc) ** 2) / (2 * scale**2)
        b = -((abs_q + loc) ** 2) / (2 * scale**2)
        log_sum = np.logaddexp(a, b)
        log_density = log_sum - np.log(np.sqrt(2 * np.pi) * scale)
        return np.where(q < 0, -np.inf, log_density)
    density = (
        1
        / (np.sqrt(2 * np.pi) * scale)
        * (
            np.exp(-((abs_q - loc) ** 2) / (2 * scale**2))
            + np.exp(-((abs_q + loc) ** 2) / (2 * scale**2))
        )
    )
    return np.where(q < 0, 0.0, density)


def pfnorm(q, loc=0, scale=1):
    """Folded Normal distribution CDF.

    Parameters
    ----------
    q : array_like
        Quantiles.
    loc : float
        Location parameter.
    scale : float
        Scale parameter.

    Returns
    -------
    array
        CDF values.
    """
    q = np.asarray(q)
    result = (
        stats.norm.cdf(q, loc=loc, scale=scale)
        + stats.norm.cdf(q, loc=-loc, scale=scale)
        - 1
    )
    result = np.where(q < 0, 0, result)
    return result


def qfnorm(p, loc=0, scale=1):
    """Folded Normal distribution quantile function.

    Parameters
    ----------
    p : array_like
        Probabilities.
    loc : float
        Location parameter.
    scale : float
        Scale parameter.

    Returns
    -------
    array
        Quantile values.
    """
    p = np.atleast_1d(p)
    result = np.zeros_like(p, dtype=float)

    for i, pi in enumerate(p):
        if pi == 0:
            result[i] = 0
        elif pi == 1:
            result[i] = np.inf
        else:

            def objective(x):
                return pfnorm(x, loc, scale) - pi

            try:
                result[i] = optimize.brentq(objective, 0, loc + 10 * scale)
            except ValueError:
                result[i] = abs(stats.norm.ppf(pi, loc=loc, scale=scale))

    return result.squeeze()


def rfnorm(n, loc=0, scale=1):
    """Folded Normal distribution random number generation.

    Parameters
    ----------
    n : int
        Number of observations.
    loc : float
        Location parameter.
    scale : float
        Scale parameter.

    Returns
    -------
    array
        Random values.
    """
    return np.abs(stats.norm.rvs(loc=loc, scale=scale, size=n))
